### PR TITLE
new rl expert (including side pick etc.)

### PR DESCRIPTION
**Pull request by Jim-Young6709**

### CODE REVIEW

File: isaacgymenvs/cfg/task/DexExpTable.yaml [hunk 1]
```diff
@@ -17,6 +17,8 @@ reward:
     target_lift_dis: 0.1
     target_quat: [1.0, 0.0, 0.0, 0.0]
     curl_reaching_threshold: 0.15
+    success_timeout: 2.0
```
Comment (Jim-Young6709) [inline isaacgymenvs/cfg/task/DexExpTable.yaml]: move this to base yaml

File: isaacgymenvs/cfg/task/DexExpTable.yaml [hunk 2]
```diff
@@ -32,3 +34,7 @@ reward:
     w_actionreg: -0.001
 
   actionreg_type: "" # "delta_joint_action", "delta_eef_action", "delta_qd"
+
+  # @ray checks where the usage is
```
Comment (Jim-Young6709) [inline isaacgymenvs/cfg/task/DexExpTable.yaml]: set to base as well

File: isaacgymenvs/cfg/task/DexExpTableSide.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    use_xy: False
+
+reward:
+  params:
+    target_pos: [0.5, 0.0, 0.3]
+    target_lift_dis: 0.1
+    target_quat_right: [-0.7071067811865475, 0.0, 0.0, 0.7071067811865476]
+    target_quat_left: [0.7071067811865475, 0.0, 0.0, 0.7071067811865476]
+    curl_reaching_threshold: 0.15
+    success_timeout: 2.0
```
Comment (Jim-Young6709) [inline isaacgymenvs/cfg/task/DexExpTableSide.yaml]: move 2 base yaml

File: isaacgymenvs/cfg/task/DexExpTableSide.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    w_hand_obj: 1.0
+    w_obj_goal: 4.0
+    w_hand_orientation: 2.0
+    w_lift: 2.0
+    w_curl: 0.5
+    w_actionreg: -0.001
+
+  actionreg_type: "" # "delta_joint_action", "delta_eef_action", "delta_qd"
+
+  # @ray checks where the usage is
```
Comment (Jim-Young6709) [inline isaacgymenvs/cfg/task/DexExpTableSide.yaml]: move 2 base yaml

File: isaacgymenvs/cfg/task/DexExpTableSide.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  eef_init:
+    limits_xyz_min: [-0.3, -0.37, -1.0]
+    limits_xyz_max: [0.15, 0.37, 1.0]
+    side_distance: 0.4
+    side_distance_noise: 0.1
+    target_quat_noise_deg: 10
+    dir_y_max: -0.8
+    dir_z_max: 0.75
+    dir_z_min: -0.25
```
Comment (Jim-Young6709) [inline isaacgymenvs/cfg/task/DexExpTableSide.yaml]: merge this with limits xyz min max

File: isaacgymenvs/tasks/franka_leap.py [hunk 6]
```diff
@@ -42,12 +42,28 @@ def __init__(self, cfg, rl_device, sim_device, graphics_device_id, headless, vir
...
         self.object_wrench_args = self.cfg["env"]["object_wrench"]
         self.eef_init = self.cfg["env"]["eef_init"]
         self.distractor_settings = self.cfg["env"]["distractor_settings"]
         self.video_logging = self.cfg["env"]["video_logging"]
         self.video_dir = os.path.join('videos', self.cfg["name"] + '_{date:%d-%H-%M-%S}'.format(date=datetime.now()))
         os.makedirs(self.video_dir, exist_ok=True)
+        
+        self.log_per_object_success = self.cfg["env"]["log_per_object_success"]["capture"]
+        self.log_per_object_success_freq = int(self.cfg["env"]["log_per_object_success"]["freq"])
+        run_name = None
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:56]: wandb

File: isaacgymenvs/tasks/franka_leap.py [hunk 7]
```diff
@@ -155,10 +175,28 @@ def _post_init_buffers(self):
...
         self.delta_joint_actions = torch.zeros((self.num_envs, self.num_robot_dofs), device=self.device, dtype=torch.float) # Current delta actions to be deployed
         self.delta_eef_actions = torch.zeros((self.num_envs, self.num_robot_dofs-1), device=self.device, dtype=torch.float) # Current delta actions to be deployed at the end effector
-        self.success_flags = torch.zeros((self.num_envs,), dtype=torch.float32, device=self.device) # 1 if success condition has been achieved at any step, 0 otherwise
+
+        # @ray per step success tracking
         self.success_5cm_per_step = torch.zeros((self.num_envs,), dtype=torch.bool, device=self.device) # success within 5cm threshold
-        self.lifting_flags = torch.zeros((self.num_envs,), dtype=torch.float32, device=self.device)
         self.lifting_5cm_per_step = torch.zeros((self.num_envs,), dtype=torch.bool, device=self.device)
+        # @ray instantaneous success tracking, true if condition met at current step
+        self.success_flags_instant = torch.zeros((self.num_envs,), dtype=torch.float32, device=self.device) # 1 if success condition has been achieved at any step, 0 otherwise
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:183]: what are the success flag vars rn

File: isaacgymenvs/tasks/franka_leap.py [hunk 8]
```diff
@@ -155,10 +175,28 @@ def _post_init_buffers(self):
...
+        self.lifting_duration = torch.zeros((self.num_envs,), dtype=torch.float32, device=self.device) 
+        self.success_long_enough = torch.zeros((self.num_envs,), dtype=torch.bool, device=self.device) # true if success duration > threshold in any part of an episode, note that this does not reset until episode end
+        self.lifting_long_enough = torch.zeros((self.num_envs,), dtype=torch.bool, device=self.device)
+        self.success_flags = torch.zeros((self.num_envs,), dtype=torch.float32, device=self.device) # 1 if success condition has been achieved during the episode, 0 otherwise
+        self.lifting_flags = torch.zeros((self.num_envs,), dtype=torch.float32, device=self.device)
+
+        # @ray 
+        # per object success rate tracking
+        # need to be post init to get num_objects
+        self.per_object_episode_counts = torch.zeros((self.num_objects,), dtype=torch.int64, device=self.device)
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:196]: self.num_objects should be define in this class

File: isaacgymenvs/tasks/franka_leap.py [hunk 9]
```diff
@@ -221,6 +259,22 @@ def _init_cuRobo_ik_solver(self):
         )
         self.ik_solver = IKSolver(ik_config)
 
+        if self.debug_viz:
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:280]: what will be visualized here?

File: isaacgymenvs/tasks/franka_leap.py [hunk 10]
```diff
@@ -363,20 +417,34 @@ def init_data(self, actor_num):
...
+        # @ray for side picks we need to switch between left and right rot targets
+        has_lr_quat = "target_quat_right" in self.cfg["reward"]["params"] and "target_quat_left" in self.cfg["reward"]["params"]
+        if has_lr_quat:
+            target_quat_right = to_torch(self.cfg["reward"]["params"]["target_quat_right"], device=self.device).unsqueeze(0).repeat(self.num_envs, 1)
+            target_quat_right_norm = torch.norm(target_quat_right, dim=1, keepdim=True)
+            target_quat_right = target_quat_right / (target_quat_right_norm + 1e-10)
+            target_quat_left = to_torch(self.cfg["reward"]["params"]["target_quat_left"], device=self.device).unsqueeze(0).repeat(self.num_envs, 1)
+            target_quat_left_norm = torch.norm(target_quat_left, dim=1, keepdim=True)
+            target_quat_left = target_quat_left / (target_quat_left_norm + 1e-10)
+            target_quat = target_quat_right.clone()
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py]: so its hardcoded to right rn?

File: isaacgymenvs/tasks/franka_leap.py [hunk 11]
```diff
@@ -398,8 +466,19 @@ def init_data(self, actor_num):
...
+                    "target_rot_6d_right": matrix_to_rotation_6d(quaternion_to_matrix_ig(target_quat_right)),
+                    "target_rot_6d_left": matrix_to_rotation_6d(quaternion_to_matrix_ig(target_quat_left)),
+                }
+                if has_lr_quat else {}
+            ),
             "curl_reaching_threshold": to_torch(self.cfg["reward"]["params"]["curl_reaching_threshold"], device=self.device),
-            "object_init_height": self.mesh_aabb_extents[:, 2] / 2 + self.table_surface_height,
+            "success_timeout": to_torch(self.cfg["reward"]["params"]["success_timeout"], device=self.device),
+            "lifting_timeout": to_torch(self.cfg["reward"]["params"]["lifting_timeout"], device=self.device),
+            "object_init_height": self.mesh_aabb_extents[:, 2] * self.object_center_z_scale + self.table_surface_height, # @ray default center is 1/2 z height
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:481]: make sure to seperate object bb center and rl grasp center

File: isaacgymenvs/tasks/franka_leap.py [hunk 12]
```diff
@@ -669,9 +788,16 @@ def _update_states(self):
         target_rot_mat_in_eef_frame = torch.matmul(eef_rot_mat.transpose(1, 2), target_rot_mat)
         target_to_eef_rot_6d = matrix_to_rotation_6d(target_rot_mat_in_eef_frame)
 
-        point_matching_err = self._get_eef_point_matching_err(
+        # @ray we don't need pos and rot error except for side grasp table
+        # probably should refactor to use separate update states later
+        point_matching_err_target = self._get_eef_point_matching_err(
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:767]: what's the differences between these two errs?

File: isaacgymenvs/tasks/franka_leap.py [hunk 13]
```diff
@@ -535,19 +617,46 @@ def _create_mesh(self, mesh_path, pos, scale, quat=[0, 0, 0, 1], fix_base_link=T
...
+        # obj_str2int = {}
+        # try:
+        #     with open(obj_mapping_path, "r") as f:
+        #         obj_str2int = json.load(f)
+        #     if not obj_str2int:
+        #         raise ValueError("Object type mapping is empty")
+        # except FileNotFoundError:
+        #     print("Object mapping file not found.")
+
+        # @ray new loading with backwards
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py]: convert mesh 69 to the new mesh file format and then remove the support for the old format

File: isaacgymenvs/tasks/franka_leap.py [hunk 14]
```diff
@@ -1003,7 +1138,7 @@ def get_joint_limits_leap(self):
         return lower_limits, upper_limits
 
     # visualization
-    def set_viewer(self, pos=[1.5, 0.0, 0.7], target=[0.5, 0.0, 0.1]):
+    def set_viewer(self, pos=[1.5, 1.0, 0.7], target=[0.5, 0.0, 0.1]):
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:1131]: what does this new angle look like

File: isaacgymenvs/tasks/franka_leap.py [hunk 15]
```diff
@@ -1089,16 +1224,32 @@ def get_camera_render(self):
     def video_logger(self):
         render_step = self.sim_steps % self.video_logging["freq"]
         if render_step == 0:
-            self.video_ims = []
+            self.video_writers = []
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:1217]: what's the new render logic here, is the 'record all object' also implemented here?

File: isaacgymenvs/tasks/franka_leap.py [hunk 16]
```diff
@@ -1107,23 +1258,66 @@ def video_logger(self):
...
+                for idx, env_idx in enumerate(self.video_env_ids):
+                    path = os.path.join(
+                        self.video_dir,
+                        f"viz_step{self.video_step_start}_env{env_idx}.mp4"
+                    )
+                    wandb.log(
+                        {f"visualization/video_env_{env_idx}": wandb.Video(path)},
+                        commit=(idx == len(self.video_env_ids) - 1),
+                    )
+        # render_step = self.sim_steps % self.video_logging["freq"]
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py]: remove comment?

File: isaacgymenvs/tasks/franka_leap.py [hunk 17]
```diff
@@ -1201,8 +1434,15 @@ def reset_idx(self, env_ids=None):
         )
 
         self.set_robot_joint_state(reset_joint_config, env_ids=env_ids)
-        self.success_flags[env_ids] = 0
-        self.lifting_flags[env_ids] = 0
+        self.success_flags_instant[env_ids] = 0
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:1411]: make sure the table top-down grasp env is compatible with all the changes here

File: isaacgymenvs/tasks/franka_leap.py (hunk 10 above)
Comment (Jim-Young6709) [reply isaacgymenvs/tasks/franka_leap.py]: rm left right, save the one to target_quat, you know what to do :-)

File: isaacgymenvs/tasks/franka_leap.py [hunk 18]
```diff
@@ -398,8 +466,19 @@ def init_data(self, actor_num):
             "target_lift_dis": target_lift_dis,
             "target_quat": target_quat,
             "target_rot_6d": matrix_to_rotation_6d(quaternion_to_matrix_ig(target_quat)),
+            **(
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py]: this will be removed

File: isaacgymenvs/tasks/franka_leap.py (hunk 13 above)
Comment (Jim-Young6709) [reply isaacgymenvs/tasks/franka_leap.py]: (just a note)

File: isaacgymenvs/tasks/franka_leap.py (hunk 11 above)
Comment (Jim-Young6709) [reply isaacgymenvs/tasks/franka_leap.py:481]: (just a note)

File: isaacgymenvs/tasks/franka_leap.py (hunk 18 above)
Comment (Jim-Young6709) [reply isaacgymenvs/tasks/franka_leap.py]: TODO

File: isaacgymenvs/tasks/franka_leap.py (hunk 10 above)
Comment (Jim-Young6709) [reply isaacgymenvs/tasks/franka_leap.py]: TODO

File: isaacgymenvs/tasks/franka_leap.py [hunk 19]
```diff
@@ -669,9 +788,16 @@ def _update_states(self):
...
-        point_matching_err = self._get_eef_point_matching_err(
+        # @ray we don't need pos and rot error except for side grasp table
+        # probably should refactor to use separate update states later
+        point_matching_err_target = self._get_eef_point_matching_err(
             curent_eef_pos7=self._eef_state[:, :7],
-            target_eef_pos7=torch.cat([self.reward_settings["target_pos"], self.reward_settings["target_quat"]], dim=-1)
+            target_eef_pos7=torch.cat([self.reward_settings["target_pos"], self.reward_settings["target_quat"]], dim=-1),
+        )
+        hand_eef_pos7_rot = torch.cat([self._eef_state[:, :3], self.reward_settings["target_quat"]], dim=-1)
+        point_matching_err_hand = self._get_eef_point_matching_err(
```
Comment (Jim-Young6709) [inline isaacgymenvs/tasks/franka_leap.py:772]: this reward / err should be conditioned on object orientation

TODO

File: isaacgymenvs/cfg/task/DexExpTableSide.yaml (hunk 5 above)
Comment (Jim-Young6709) [reply isaacgymenvs/cfg/task/DexExpTableSide.yaml]: TODO

File: isaacgymenvs/tasks/franka_leap.py (hunk 13 above)
Comment (DaedalusDelta) [reply isaacgymenvs/tasks/franka_leap.py]: Created conversion script "convert_legacy_meshes.py". Now legacy format is removed and no longer accepted.

File: isaacgymenvs/cfg/task/DexExpTableSide.yaml (hunk 5 above)
Comment (DaedalusDelta) [reply isaacgymenvs/cfg/task/DexExpTableSide.yaml]: removed dir limits to only use limits xyz min max.

File: isaacgymenvs/tasks/franka_leap.py (hunk 11 above)
Comment (DaedalusDelta) [reply isaacgymenvs/tasks/franka_leap.py:481]: done, this is only applied to the side grasp policy. The rest of the envs still use the object bb center for reward.

File: isaacgymenvs/tasks/franka_leap.py (hunk 18 above)
Comment (DaedalusDelta) [reply isaacgymenvs/tasks/franka_leap.py]: removed, now all task envs use target_quat. side grasp envs keep a self.target_quat_right, self.target_quat_left internally, update reward_settings["target_quat"] in reset_idx to maximally share reward_settings.

File: isaacgymenvs/tasks/franka_leap.py (hunk 19 above)
Comment (DaedalusDelta) [reply isaacgymenvs/tasks/franka_leap.py:772]: TBD

File: isaacgymenvs/tasks/franka_leap.py (hunk 10 above)
Comment (DaedalusDelta) [reply isaacgymenvs/tasks/franka_leap.py]: remvoed, only target_quat is used now.
